Source/core/MessageDispatcher.h

```diff
@@ -0,0 +1,422 @@
+        *
+        * @param length length of message
+        * @param value buffer 
+        * @return uint32_t ERROR_WRITE_ERROR: failed to reserve enough space - eg, value size is exceeding max cyclic buffer size
+        *                  ERROR_NONE: OK
+        */
+        uint32_t PushData(const uint16_t length, const uint8_t* value)
+        {
+            _dataLock.Lock();
+
+            ASSERT(length > 0);
+            uint32_t result = Core::ERROR_WRITE_ERROR;
+            const uint16_t fullLength = sizeof(length) + length; // headerLength + informationLength
+
+            const uint16_t reservedLength = _dataBuffer.Reserve(fullLength);
+
+            if (reservedLength >= fullLength) {
+                //no need to serialize because we can write to CyclicBuffer step by step
+                _dataBuffer.Write(reinterpret_cast<const uint8_t*>(&fullLength), sizeof(fullLength)); //fullLength
+                _dataBuffer.Write(value, length); //value
+                result = Core::ERROR_NONE;
+
+            } else {
+                TRACE_L1("Buffer to small to fit message!\n");
+            }
+
+            _dataLock.Unlock();
+
+            return result;
+        }
+
+        /**
+         * @brief Read data after doorbell ringed. The buffer size should be at least of size of message + sizeof(uint16_t). 
+         *        If buffer is too small to fit whole message it will be partially filled. 
+         *        When reading, one should then skip last two bytes of passed buffer
+         * 
+         * @param outLength ERROR_NONE - read bytes. 
+         *                  ERROR_GENERAL - mimimal required bytes to fit whole message.
+         *                  ERROR_READ_ERROR - the same value as passed in                       
+         * @param outValue buffer - in case of ERROR_GENERAL: last two bytes should be skipped                
+         * @return uint32_t ERROR_READ_ERROR - unable to read or data is corrupted
+         *                  ERROR_NONE - OK
+         *                  ERROR_GENERAL - buffer too small to fit whole message at once
+         */
+        uint32_t PopData(uint16_t& outLength, uint8_t* outValue)
+        {
+            ASSERT(_dataBuffer.IsValid());
+
+            _dataLock.Lock();
+            uint32_t result = Core::ERROR_READ_ERROR;
+
+            if (_dataBuffer.Validate()) {
+
+                uint32_t length = _dataBuffer.Read(outValue, outLength, true);
+
+                //did not even receive length of the full message
+                if (length < 2) {
+                    TRACE_L1("Inconsistent message\n");
+                    _dataBuffer.Flush();
+                } else if (length > outLength) {
+                    TRACE_L1("Lost part of the message\n");
+                    result = Core::ERROR_GENERAL;
+
+                    //first two bytes are sizeof full messsage.
+                    std::copy_n(outValue + sizeof(outLength), outLength - sizeof(outLength), outValue);
```

Comment:
Suggest to adapt this from the ReadSize() using the cursor from the Cyclic buffer perspective to avoid this "memcpy". Lets discuss this.